File: src/utils/pdfGenerator.ts
import jsPDF from 'jspdf';
import type { ScanResult } from './urlAnalyzer';

export function generatePDF(result: ScanResult): jsPDF {
  const doc = new jsPDF();
  const pageWidth = doc.internal.pageSize.getWidth();
  
  // Header
  doc.setFillColor(10, 14, 26);
  doc.rect(0, 0, pageWidth, 50, 'F');
  
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(24);
  doc.setFont('helvetica', 'bold');
  doc.text('PhishGuard AI', 20, 25);
  
  doc.setTextColor(255, 255, 255);
  doc.setFontSize(12);
  doc.setFont('helvetica', 'normal');
  doc.text('Phishing Detection Report', 20, 38);
  
  // Report metadata
  doc.setTextColor(100, 100, 100);
  doc.setFontSize(10);
  doc.text(`Generated: ${new Date(result.timestamp).toLocaleString()}`, pageWidth - 20, 20, { align: 'right' });
  doc.text(`Report ID: ${result.id}`, pageWidth - 20, 28, { align: 'right' });
  
  let y = 65;
  
  // URL Section
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(14);
  doc.setFont('helvetica', 'bold');
  doc.text('Target URL', 20, y);
  y += 10;
  
  doc.setFillColor(240, 240, 240);
  doc.rect(20, y - 5, pageWidth - 40, 12, 'F');
  doc.setTextColor(50, 50, 50);
  doc.setFontSize(10);
  doc.setFont('helvetica', 'normal');
  doc.text(result.url, 25, y + 2);
  y += 20;
  
  // Risk Score
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(14);
  doc.setFont('helvetica', 'bold');
  doc.text('Risk Assessment', 20, y);
  y += 10;
  
  const scoreColor = result.riskScore <= 20 ? [0, 255, 136] : 
                     result.riskScore <= 40 ? [0, 212, 255] :
                     result.riskScore <= 60 ? [255, 221, 0] :
                     result.riskScore <= 80 ? [255, 170, 0] : [255, 51, 102];
  
  doc.setFillColor(scoreColor[0], scoreColor[1], scoreColor[2]);
  doc.roundedRect(20, y - 5, 60, 25, 3, 3, 'F');
  doc.setTextColor(255, 255, 255);
  doc.setFontSize(20);
  doc.setFont('helvetica', 'bold');
  doc.text(`${result.riskScore}`, 35, y + 12);
  doc.setFontSize(10);
  doc.text('/ 100', 55, y + 12);
  
  doc.setTextColor(50, 50, 50);
  doc.setFontSize(14);
  doc.setFont('helvetica', 'bold');
  doc.text(result.category, 90, y + 8);
  y += 35;
  
  // Features
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(14);
  doc.text('URL Analysis Features', 20, y);
  y += 10;
  
  const features = [
    ['URL Length', `${result.features.urlLength} characters`],
    ['Dot Count', `${result.features.dotCount}`],
    ['Hyphen Count', `${result.features.hyphenCount}`],
    ['Has @ Symbol', result.features.hasAtSymbol ? 'Yes' : 'No'],
    ['HTTPS Available', result.features.hasHTTPS ? 'Yes' : 'No'],
    ['SSL Valid', result.features.sslValid ? 'Yes' : 'No'],
    ['Domain Age', `${result.features.domainAge} days`],
    ['Domain Expiry', `${result.features.domainExpiry} days`],
    ['Subdomain Count', `${result.features.subdomainCount}`],
    ['Redirect Count', `${result.features.redirectCount}`],
    ['IP Address', result.features.ipAddress ? 'Yes' : 'No'],
    ['Has Port', result.features.hasPort ? 'Yes' : 'No'],
  ];
  
  doc.setFontSize(10);
  features.forEach(([label, value], i) => {
    const rowY = y + (i * 8);
    if (rowY > 270) {
      doc.addPage();
      y = 20;
    }
    doc.setTextColor(100, 100, 100);
    doc.setFont('helvetica', 'normal');
    doc.text(label, 20, rowY);
    doc.setTextColor(50, 50, 50);
    doc.setFont('helvetica', 'bold');
    doc.text(value, 80, rowY);
  });
  y += features.length * 8 + 15;
  
  // VirusTotal
  if (y > 240) {
    doc.addPage();
    y = 20;
  }
  
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(14);
  doc.setFont('helvetica', 'bold');
  doc.text('VirusTotal Intelligence', 20, y);
  y += 10;
  
  doc.setFontSize(10);
  doc.setTextColor(100, 100, 100);
  doc.setFont('helvetica', 'normal');
  doc.text('Reputation Score:', 20, y);
  doc.setTextColor(50, 50, 50);
  doc.setFont('helvetica', 'bold');
  doc.text(`${result.virusTotal.reputation}/100`, 80, y);
  y += 8;
  
  doc.setTextColor(100, 100, 100);
  doc.setFont('helvetica', 'normal');
  doc.text('Detections:', 20, y);
  doc.setTextColor(50, 50, 50);
  doc.setFont('helvetica', 'bold');
  doc.text(`${result.virusTotal.detections}/${result.virusTotal.totalEngines} engines`, 80, y);
  y += 12;
  
  // WHOIS
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(14);
  doc.setFont('helvetica', 'bold');
  doc.text('WHOIS Information', 20, y);
  y += 10;
  
  const whoisData = [
    ['Registrar', result.whois.registrar],
    ['Created', result.whois.created],
    ['Expires', result.whois.expires],
    ['Updated', result.whois.updated],
    ['Country', result.whois.country],
  ];
  
  doc.setFontSize(10);
  whoisData.forEach(([label, value]) => {
    doc.setTextColor(100, 100, 100);
    doc.setFont('helvetica', 'normal');
    doc.text(label, 20, y);
    doc.setTextColor(50, 50, 50);
    doc.setFont('helvetica', 'bold');
    doc.text(value, 80, y);
    y += 8;
  });
  y += 10;
  
  // SSL
  if (y > 240) {
    doc.addPage();
    y = 20;
  }
  
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(14);
  doc.setFont('helvetica', 'bold');
  doc.text('SSL Certificate', 20, y);
  y += 10;
  
  const sslData = [
    ['Issuer', result.ssl.issuer],
    ['Valid From', result.ssl.validFrom],
    ['Valid To', result.ssl.validTo],
    ['Cipher', result.ssl.cipher],
    ['Protocol', result.ssl.protocol],
  ];
  
  doc.setFontSize(10);
  sslData.forEach(([label, value]) => {
    doc.setTextColor(100, 100, 100);
    doc.setFont('helvetica', 'normal');
    doc.text(label, 20, y);
    doc.setTextColor(50, 50, 50);
    doc.setFont('helvetica', 'bold');
    doc.text(value, 80, y);
    y += 8;
  });
  y += 15;
  
  // Recommendations
  if (y > 220) {
    doc.addPage();
    y = 20;
  }
  
  doc.setTextColor(0, 212, 255);
  doc.setFontSize(14);
  doc.setFont('helvetica', 'bold');
  doc.text('Recommendations', 20, y);
  y += 10;
  
  doc.setFontSize(10);
  result.recommendations.forEach((rec, i) => {
    doc.setTextColor(50, 50, 50);
    doc.setFont('helvetica', 'normal');
    const lines = doc.splitTextToSize(`${i + 1}. ${rec}`, pageWidth - 50);
    doc.text(lines, 20, y);
    y += lines.length * 6 + 4;
  });
  
  // Footer
  const totalPages = doc.getNumberOfPages();
  for (let i = 1; i <= totalPages; i++) {
    doc.setPage(i);
    doc.setFillColor(10, 14, 26);
    doc.rect(0, doc.internal.pageSize.getHeight() - 15, pageWidth, 15, 'F');
    doc.setTextColor(150, 150, 150);
    doc.setFontSize(8);
    doc.setFont('helvetica', 'normal');
    doc.text(`PhishGuard AI - Page ${i} of ${totalPages}`, pageWidth / 2, doc.internal.pageSize.getHeight() - 6, { align: 'center' });
  }
  
  return doc;
}

export function downloadPDF(result: ScanResult) {
  const doc = generatePDF(result);
  doc.save(`phishguard-report-${result.id}.pdf`);
}

File: utils/pdf_generator.py
"""
PhishGuard AI - PDF Report Generator
Generate professional PDF reports for scan results
"""

from reportlab.lib.pagesizes import A4
from reportlab.lib import colors
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle, PageBreak
from reportlab.lib.units import inch
from reportlab.pdfgen import canvas
from datetime import datetime
import io
import base64

def generate_pdf_report(scan_data):
    """Generate PDF report from scan data"""
    buffer = io.BytesIO()
    doc = SimpleDocTemplate(buffer, pagesize=A4, topMargin=0.5*inch, bottomMargin=0.5*inch)
    styles = getSampleStyleSheet()
    elements = []
    
    # Custom styles
    title_style = ParagraphStyle(
        'CustomTitle',
        parent=styles['Heading1'],
        fontSize=24,
        textColor=colors.HexColor('#00d4ff'),
        spaceAfter=30,
        alignment=1
    )
    
    heading_style = ParagraphStyle(
        'CustomHeading',
        parent=styles['Heading2'],
        fontSize=16,
        textColor=colors.HexColor('#00d4ff'),
        spaceAfter=12,
        spaceBefore=20
    )
    
    # Title
    elements.append(Paragraph("PhishGuard AI", title_style))
    elements.append(Paragraph("Phishing Detection Report", styles['Heading2']))
    elements.append(Spacer(1, 20))
    
    # Report metadata
    report_date = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    elements.append(Paragraph(f"<b>Report ID:</b> {scan_data.get('id', 'N/A')}", styles['Normal']))
    elements.append(Paragraph(f"<b>Generated:</b> {report_date}", styles['Normal']))
    elements.append(Spacer(1, 20))
    
    # URL Section
    elements.append(Paragraph("Target URL", heading_style))
    elements.append(Paragraph(f"<font color='#0066cc'>{scan_data.get('url', 'N/A')}</font>", styles['Normal']))
    elements.append(Spacer(1, 20))
    
    # Risk Assessment
    elements.append(Paragraph("Risk Assessment", heading_style))
    risk_score = scan_data.get('risk_score', 0)
    category = scan_data.get('category', 'Unknown')
    
    risk_color = '#00ff88' if risk_score <= 20 else '#00d4ff' if risk_score <= 40 else '#ffdd00' if risk_score <= 60 else '#ffaa00' if risk_score <= 80 else '#ff3366'
    
    risk_data = [
        ['Risk Score', f"{risk_score}/100"],
        ['Category', category],
        ['Status', 'SAFE' if risk_score <= 40 else 'WARNING' if risk_score <= 60 else 'DANGEROUS']
    ]
    
    risk_table = Table(risk_data, colWidths=[2*inch, 3*inch])
    risk_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (0, -1), colors.HexColor('#f0f0f0')),
        ('TEXTCOLOR', (0, 0), (-1, -1), colors.black),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('FONTNAME', (0, 0), (0, -1), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, -1), 11),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 8),
        ('TOPPADDING', (0, 0), (-1, -1), 8),
        ('GRID', (0, 0), (-1, -1), 1, colors.grey)
    ]))
    
    elements.append(risk_table)
    elements.append(Spacer(1, 20))
    
    # URL Analysis Features
    elements.append(Paragraph("URL Analysis Features", heading_style))
    features = scan_data.get('features', {})
    
    features_data = []
    feature_mapping = [
        ('url_length', 'URL Length'),
        ('dot_count', 'Dot Count'),
        ('hyphen_count', 'Hyphen Count'),
        ('has_at', 'Has @ Symbol'),
        ('has_https', 'HTTPS Available'),
        ('path_length', 'Path Length'),
        ('query_length', 'Query Length'),
        ('subdomain_count', 'Subdomain Count'),
        ('ip_address', 'IP Address'),
        ('tiny_url', 'Tiny URL'),
        ('has_port', 'Has Port'),
        ('suspicious_keywords', 'Suspicious Keywords'),
        ('brand_impersonation', 'Brand Impersonation')
    ]
    
    for key, label in feature_mapping:
        if key in features:
            value = features[key]
            if isinstance(value, bool):
                value = 'Yes' if value else 'No'
            features_data.append([label, str(value)])
    
    if features_data:
        features_table = Table(features_data, colWidths=[2.5*inch, 2.5*inch])
        features_table.setStyle(TableStyle([
            ('BACKGROUND', (0, 0), (-1, 0), colors.HexColor('#e0e0e0')),
            ('TEXTCOLOR', (0, 0), (-1, -1), colors.black),
            ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
            ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
            ('FONTSIZE', (0, 0), (-1, -1), 10),
            ('BOTTOMPADDING', (0, 0), (-1, -1), 6),
            ('TOPPADDING', (0, 0), (-1, -1), 6),
            ('GRID', (0, 0), (-1, -1), 1, colors.grey)
        ]))
        elements.append(features_table)
        elements.append(Spacer(1, 20))
    
    # VirusTotal Intelligence
    elements.append(Paragraph("VirusTotal Intelligence", heading_style))
    vt = scan_data.get('virus_total', {})
    
    vt_data = [
        ['Reputation Score', f"{vt.get('reputation', 0)}/100"],
        ['Detections', f"{vt.get('detections', 0)}/{vt.get('total_engines', 0)}"],
        ['Status', 'Safe' if vt.get('detections', 0) == 0 else 'Flagged']
    ]
    
    vt_table = Table(vt_data, colWidths=[2*inch, 3*inch])
    vt_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (0, -1), colors.HexColor('#f0f0f0')),
        ('TEXTCOLOR', (0, 0), (-1, -1), colors.black),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('FONTNAME', (0, 0), (0, -1), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, -1), 11),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 8),
        ('TOPPADDING', (0, 0), (-1, -1), 8),
        ('GRID', (0, 0), (-1, -1), 1, colors.grey)
    ]))
    
    elements.append(vt_table)
    elements.append(Spacer(1, 20))
    
    # WHOIS Information
    elements.append(Paragraph("WHOIS Information", heading_style))
    whois = scan_data.get('whois', {})
    
    whois_data = [
        ['Registrar', str(whois.get('registrar', 'N/A'))],
        ['Creation Date', str(whois.get('creation_date', 'N/A'))],
        ['Expiration Date', str(whois.get('expiration_date', 'N/A'))],
        ['Country', str(whois.get('country', 'N/A'))]
    ]
    
    whois_table = Table(whois_data, colWidths=[2*inch, 3*inch])
    whois_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (0, -1), colors.HexColor('#f0f0f0')),
        ('TEXTCOLOR', (0, 0), (-1, -1), colors.black),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('FONTNAME', (0, 0), (0, -1), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, -1), 11),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 8),
        ('TOPPADDING', (0, 0), (-1, -1), 8),
        ('GRID', (0, 0), (-1, -1), 1, colors.grey)
    ]))
    
    elements.append(whois_table)
    elements.append(Spacer(1, 20))
    
    # SSL Certificate
    elements.append(Paragraph("SSL Certificate", heading_style))
    ssl_info = scan_data.get('ssl_info', {})
    
    ssl_data = [
        ['Valid', 'Yes' if ssl_info.get('valid') else 'No'],
        ['Issuer', str(ssl_info.get('issuer', 'N/A'))],
        ['Valid From', str(ssl_info.get('valid_from', 'N/A'))],
        ['Valid To', str(ssl_info.get('valid_to', 'N/A'))],
        ['Days Until Expiry', str(ssl_info.get('days_until_expiry', 'N/A'))]
    ]
    
    ssl_table = Table(ssl_data, colWidths=[2*inch, 3*inch])
    ssl_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (0, -1), colors.HexColor('#f0f0f0')),
        ('TEXTCOLOR', (0, 0), (-1, -1), colors.black),
        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),
        ('FONTNAME', (0, 0), (0, -1), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, -1), 11),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 8),
        ('TOPPADDING', (0, 0), (-1, -1), 8),
        ('GRID', (0, 0), (-1, -1), 1, colors.grey)
    ]))
    
    elements.append(ssl_table)
    elements.append(Spacer(1, 20))
    
    # Recommendations
    elements.append(Paragraph("Recommendations", heading_style))
    recommendations = scan_data.get('recommendations', [])
    
    for i, rec in enumerate(recommendations, 1):
        elements.append(Paragraph(f"{i}. {rec}", styles['Normal']))
    
    elements.append(Spacer(1, 30))
    
    # Footer
    elements.append(Spacer(1, 20))
    elements.append(Paragraph("<i>Generated by PhishGuard AI - Advanced Phishing Detection Platform</i>", styles['Italic']))
    
    # Build PDF
    doc.build(elements)
    
    # Get PDF bytes
    pdf_bytes = buffer.getvalue()
    buffer.close()
    
    # Convert to base64
    pdf_base64 = base64.b64encode(pdf_bytes).decode('utf-8')
    
    return pdf_base64


def save_pdf_report(scan_data, filename):
    """Save PDF report to file"""
    pdf_base64 = generate_pdf_report(scan_data)
    pdf_bytes = base64.b64decode(pdf_base64)
    
    with open(filename, 'wb') as f:
        f.write(pdf_bytes)
    
    return filename
